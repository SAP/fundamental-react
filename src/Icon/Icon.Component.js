import { Icon } from '../';
import { listOfIcons } from '../utils/listOfIcons';
import path from 'path';
import React from 'react';
import { ComponentPage, Example } from '../_playground';

export const IconComponent = () => {
    let icons = (
        <div>
            {listOfIcons.map((icon, index) => {
                return (
                    <div className='demo-icon-wrapper' key={index}>
                        <Icon glyph={icon} size='xl' />
                        <h5>sap-icon--{icon}</h5>
                    </div>
                );
            })}
        </div>
    );

    return (
        <ComponentPage
            description={`Icons are used throughout the UI to save space, allow for visual clarity
                and focus, and for fun. Icons can be used adaptively if desired, but at
                this point they are used more as visual elements within other
                components.`}
            sourceModulePath={path.join(__dirname, './Icon')}
            title='Icon'>

            <Example
                centered
                title='Sizes'>
                <Icon glyph='cart' size='s' />
                <Icon glyph='cart' />
                <Icon glyph='cart' size='m' />
                <Icon glyph='cart' size='l' />
                <Icon glyph='cart' size='xl' />
            </Example>

            <Example
                centered
                omitCodeSample
                title='Available Icons'>
                {icons}
            </Example>

        </ComponentPage>
    );
};
