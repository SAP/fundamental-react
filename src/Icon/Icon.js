import classnamesBind from 'classnames/bind';
import { ICON_SIZES } from '../utils/constants';
import PropTypes from 'prop-types';
import React from 'react';
import withStyles from '../utils/withStyles';
import styles from 'fundamental-styles/dist/icon.css';

const classnames = classnamesBind.bind(styles);

/** Icons are used throughout the UI to save space, allow for visual clarity
and focus, and for fun. Icons can be used adaptively if desired, but at
this point they are used more as visual elements within other
components. */
const Icon = React.forwardRef(({ ariaHidden, ariaLabel, glyph, size, className, cssNamespace, ...props }, ref) => {

    const iconClasses = classnames(
        {
            [`sap-icon--${glyph}`]: !!glyph
        },
        className
    );

    let style;
    if (size) {
        style = { fontSize: ICON_SIZES[size] };
    }

    return (
        <i
            {...props}
            aria-hidden={ariaHidden}
            aria-label={ariaLabel}
            className={iconClasses}
            ref={ref}
            role='img'
            style={style} />
    );
});

Icon.displayName = 'Icon';

Icon.propTypes = {
    /** The icon to include. See the icon page for the list of icons */
    glyph: PropTypes.string.isRequired,
    /** The value for the `aria-hidden` attribute. Set to `true` when there is already visible text explaining the icon, or the icon is purely for decoration. */
    ariaHidden: PropTypes.bool,
    /** Text to describe the icon. The value is placed in the `aria-label` attribute. */

    ariaLabel: (props) => {
        if (
            (!props.ariaHidden && !props.ariaLabel) ||
            (!props.ariaHidden && typeof props.ariaLabel !== 'string')
        ) {
            return new Error('ariaLabel is required if ariaHidden is false and must be a string');
        }
    },
    /** CSS class(es) to add to the element */
    className: PropTypes.string,
    /** Size of the component: 's', 'm', 'l', 'xl' */
    size: PropTypes.oneOf(Object.keys(ICON_SIZES))
};

export default withStyles(Icon);
