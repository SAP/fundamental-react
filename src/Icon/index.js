export { default as Icon } from './Icon.js';
