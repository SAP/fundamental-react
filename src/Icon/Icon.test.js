import Icon from './Icon';
import { mount } from 'enzyme';
import React from 'react';

describe('<Icon />', () => {
    const mockOnClick = jest.fn();

    const defaultProps = { ariaHidden: true, glyph: 'cart' };
    const defaultIcon = <Icon {...defaultProps} />;
    const clickIcon = <Icon {...defaultProps} onClick={mockOnClick} />;
    const iconWithSize = <Icon {...defaultProps} size='s' />;
    const ariaLabelIcon = <Icon ariaLabel='Add to cart' glyph='cart' />;
    const propSpreadIcon = <Icon {...defaultProps} data-sample='Sample' />;

    test('click on icon', () => {
        let wrapper = mount(clickIcon);
        wrapper.find('.sap-icon--cart').simulate('click');
        expect(wrapper.prop('onClick')).toBeCalledTimes(1);

        wrapper = mount(iconWithSize);
        wrapper.find('.sap-icon--cart').simulate('click');
        expect(wrapper.prop('onClick')).toBeUndefined;
    });

    test('passes aria-label prop', () => {
        const element = mount(ariaLabelIcon);
        expect(element.getDOMNode().getAttribute('aria-label')).toBe('Add to cart');
    });

    test('passes the aria-hidden prop', () => {
        const element = mount(defaultIcon);
        expect(element.getDOMNode().getAttribute('aria-hidden')).toBe('true');
    });

    describe('Prop spreading', () => {
        test('should allow props to be spread to the Icon component', () => {
            const element = mount(propSpreadIcon);

            expect(
                element.getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });
    });

    test('forwards the ref', () => {
        let ref;
        class Test extends React.Component {
            constructor(props) {
                super(props);
                ref = React.createRef();
            }
            render = () => <Icon {...defaultProps} ref={ref} />;
        }
        mount(<Test />);
        expect(ref.current.tagName).toEqual('I');
        expect(ref.current.className).toEqual('sap-icon--cart');
    });
});
