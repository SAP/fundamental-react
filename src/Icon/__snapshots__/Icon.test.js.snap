// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Icon /> create icon 1`] = `
<span
  className="sap-icon--cart blue"
  onClick={[MockFunction]}
/>
`;

exports[`<Icon /> create icon 2`] = `
<span
  className="sap-icon--cart sap-icon--s"
  onClick={[Function]}
/>
`;
